core/src/main/java/org/infinispan/expiration/ExpirationManager.java

```diff
@@ -35,11 +35,32 @@
    boolean isEnabled();
 
    /**
-    * This should be invoked passing in an entry that is now expired.  Note this method assumes to not have the
-    * lock for this key and will attempt to acquire it.
+    * This should be invoked passing in an entry that is now expired.  This method assumes the lock is not acquire
+    * and may attempt to lock this key to preserve atomicity.
     * @param entry entry that is now expired
     */
    void handleInMemoryExpiration(InternalCacheEntry<K, V> entry);
 
+   /**
+    * This is to be invoked when a store entry expires.  This method assumes the lock is not acquire and may attempt
+    * to lock this key to preserve atomicity.
+    * <p>
+    * Note this method doesn't currently take a {@link InternalCacheEntry} and this is due to a limitation in the
+    * cache store API.  This may cause some values to be removed if they were updated at the same time.
```

Comment:
I wonder if we could do something to minimize the chance of this happening without an API change... at least in `SingleFileStore.purge()`, it can take a long time between the entry being added to the `entriesToPurge` collection and the listener being invoked.